storage: fix deadlock in consistency queue

tbg opened the pull request:
When `CheckConsistency` returns an error, the queue checks whether the
store is draining to decide whether the error is worth logging.

Unfortunately this check was incorrect and would block until the store
actually started draining.

A toy example of this problem is below (this will deadlock). The dual
return form of chan receive isn't non-blocking -- the second parameter
indicates whether the received value corresponds to a closing of the
channel.

Switch to a `select` instead.

```go
package main

import (
	"fmt"
)

func main() {
	ch := make(chan struct{})
	_, ok := <-ch
	fmt.Println(ok)
}
```

Touches #21824.

Release note (bug fix): Prevent the consistency checker from
deadlocking. This would previously manifest itself as a steady number of
replicas queued for consistency checking on one or more nodes and would
resolve by restarting the affected nodes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25456)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/consistency_queue.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25456#-LCRMhVO2aF2Q68YztYy:-LCRMhVO2aF2Q68YztYz:b-fi0eov) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4777beb4b64bdfbe2e850bceaf87fc59d709fee/pkg/storage/consistency_queue.go#L107)):*
> ```Go
> 
> 	// We only need this variable in the `pErr != nil` branch below, but for
> 	// better testing coverage it was moved here.
> ```

Isn't this too soon? `repl.CheckConsistency` takes some time and we want to check for quiescence after it's done. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25456)*
<!-- Sent from Reviewable.io -->

a-robinson commented on pkg/storage/consistency_queue.go:
```diff
@@ -102,19 +102,30 @@ func (q *consistencyQueue) process(
+
+	// Update the last processed time for this queue. We do this before actually
+	// processing because this queue will do it even in the case of an error
+	// (since a failed processing run could be caused by down followers).
```
It's not clear from the comment why this is worth doing specifically for the consistency queue but not for any other queues.

tschottdorf replied:
I'm really just backfilling rationale here, but this is the previous behavior. I'm going to add a bit of doubt to my comment to not make it seem like this was a decision prompted by an observable problem.

a-robinson commented on pkg/storage/replica_test.go:
```diff
@@ -9267,6 +9267,34 @@ func TestReplicaRecomputeStats(t *testing.T) {
 	}
 }
 
+// TestConsistenctQueueErrorFromCheckConsistency exercises the case in which
```
s/Consistenct/Consistency

a-robinson commented on pkg/storage/replica_test.go:
```diff
@@ -9267,6 +9267,34 @@ func TestReplicaRecomputeStats(t *testing.T) {
 
+// TestConsistenctQueueErrorFromCheckConsistency exercises the case in which
+// the queue receives an error from CheckConsistency.
+func TestConsistenctQueueErrorFromCheckConsistency(t *testing.T) {
```
Missing a leak-checker call

a-robinson reviewed (approved):
LGTM, just nits. Thanks for tracking this down.

tbg committed:
storage: fix deadlock in consistency queue

When `CheckConsistency` returns an error, the queue checks whether the
store is draining to decide whether the error is worth logging.

Unfortunately this check was incorrect and would block until the store
actually started draining.

A toy example of this problem is below (this will deadlock). The dual
return form of chan receive isn't non-blocking -- the second parameter
indicates whether the received value corresponds to a closing of the
channel.

Switch to a `select` instead.

```
package main

import (
	"fmt"
)

func main() {
	ch := make(chan struct{})
	_, ok := <-ch
	fmt.Println(ok)
}
```

Touches #21824.

Release note (bug fix): Prevent the consistency checker from
deadlocking. This would previously manifest itself as a steady number of
replicas queued for consistency checking on one or more nodes and would
resolve by restarting the affected nodes.

tschottdorf commented:
Updated the comment.

bors r=bdarnell,a-robinson
